base/multidimensional.jl

```diff
@@ -384,8 +394,10 @@ end
         @nexprs $N d->(I_d = I[d])
         idxlens = @ncall $N index_lengths A I
         @ncall $N setindex_shape_check X (d->idxlens[d])
+        J = @ncall $N decolon A I
+        @nexprs $N d->(J_d = J[d])
         Xs = start(X)
-        @nloops $N i d->(1:idxlens[d]) d->(@inbounds j_d = I_d[i_d]) begin
+        @nloops $N j d->J_d begin
```

Comment:
Nice.  This means that we once again iterate over the indices, yes?  I think that's the right way to go — it definitely will speed this up for non-linear-fast index types.  And it has a very nice path towards speeding up logical indexing with an `IndexSet` (nee `IntSet` wrapper type).  But that'll take some more thought for arbitrarily-offset indices.

It does have the downside of putting another roadblock back into https://github.com/JuliaLang/julia/issues/7903.  Perhaps `to_index` could wrap numbers with an iterable wrapper, but that also requires detangling `to_scalar_index` and `to_nonscalar_index`.